apps/src/applab/firebaseCounters.js

```diff
@@ -69,6 +69,45 @@ export function updateTableCounters(tableName, rowCountChange, updateNextId) {
 }
 
 /**
+ * Increment or reset the rate-limiting counters associated with the interval.
+ * @param {number} maxWriteCount
+ * @param {string} interval
+ * @param {number} currentTimeMs
+ * @returns {Promise}
+ */
+function incrementIntervalCounters(maxWriteCount, interval, currentTimeMs) {
+  const limitRef = getDatabase(Applab.channelId).child(`counters/limits/${interval}`);
+  return limitRef.transaction(limitData => {
+    limitData = limitData || {};
+    limitData.lastResetTime = limitData.lastResetTime || 0;
+    limitData.writeCount = (limitData.writeCount || 0) + 1;
+    if (limitData.writeCount <= maxWriteCount) {
+      return limitData;
+    } else if (limitData.lastResetTime + interval * 1000 < currentTimeMs) {
+      // The maximum number of writes has been exceeded in more than `interval` seconds.
+      // Reset the counters.
+      limitData.writeCount = 1;
+      limitData.lastResetTime = Firebase.ServerValue.TIMESTAMP;
+      return limitData;
+    } else {
+      // The maximum number of writes has been exceeded in less than `interval` seconds.
+      // Abort the transaction.
+      return;
+    }
+  }).then(transactionData => {
+    if (!transactionData.committed) {
```

Comment:
In the abort case above, you just `return;` - doesn't that mean `transactionData` would be undefined, and this would throw an error?  But it looks like you've tested this pretty thoroughly, so maybe there's a `Promise` behavior I don't understand here.